### ISSUE TITLE
Won't launch Steamlink

### ISSUE THREAD

**Issue by briantogo** [swetoast/steamlink-launcher#61]
Greetings,
I've loaded the latest release of OSMC on a PI4.  The case is an Argon One, and I installed the OS on a USB SSD.  I've installed the plugin.program.steamlink.v0.0.13.zip, but cannot open the program "Steamlink Launcher" the Add-ons menu.  When I choose "Open", I get the following error;
culr:  (3) URL using bad/illegal format or missing URL
Couldn't download archive, aborting update

I also noticed a configure option is missing in the launcher.  Starting the program from the Programs section in OSMC also gives that error.

Thanks,
Brian

**Comment by swetoast** [comment]
greetings, can you manually install the deb package steam and try and install it manually from console via ssh

```bash
    curl -o /tmp/steamlink.deb -#Of http://media.steampowered.com/steamlink/rpi/latest/steamlink.deb
    sudo dpkg -i /tmp/steamlink.deb
    rm -f /tmp/steamlink.deb
 ```
 
 im curious to find out if the current deb package is broken or not but noone seems to be able to help out troubleshooting

**Comment by swetoast** [comment]
if you get an error when running the deb package thats made by valve then close this ticket and go to steam forums and report that deb package is broken.

this small and by small launcher is really a script that only download and run the steamlink if download is ok then run if something isnt working then often its steamlink thats the main issue

**Comment by briantogo** [comment]
Thanks Swetoast, I tried out the steps above, and it seemed to install without error messages.  However, there are no signs of it anywhere within the OSMC gui.  When I send the command "steamlink", it comes back with the same error;
culr: (3) URL using bad/illegal format or missing URL
Couldn't download archive, aborting update

I noticed it fails trying to get an update.  Can I bypass the update part and just open Steamlink without the updates?

I have written Steam about the issue, and will let you know when they respond.

**Comment by swetoast** [comment]
thanks for the help i hope it resolves the issue

**Comment by swetoast** [comment]
leaving this issue open so you can report back your progress

**Comment by 0chroma** [comment]
I managed to install steamlink, but I'm still getting this error when launching the application:
![Screenshot From 2024-12-07 00-36-26](https://github.com/user-attachments/assets/ca2565d3-032a-45a0-89bc-3f426a6b9d85)

**Comment by swetoast** [comment]
yeah, all stated above go report the issue to steam on their forums

**Comment by swetoast** [comment]
would say this has something to do with the mess

https://www.raspberrypi.com/news/valves-steam-link-on-raspberry-pi/

so its in the raspbian repo and OSMC is not raspbian :+1: 

so unless the package has been change this little fun launcher might be null and void

**Comment by swetoast** [comment]
[steamlink_1.0.13_arm64.deb](https://archive.raspberrypi.org/debian/pool/main/s/steamlink/steamlink_1.0.13_arm64.deb)	2024-11-12 16:57 	27K	 
[steamlink_1.0.13_armhf.deb](https://archive.raspberrypi.org/debian/pool/main/s/steamlink/steamlink_1.0.13_armhf.deb)	2024-11-12 16:57 	28K 

anyone up for being a guinea pig ?
these are the new raspbian packages for steamlink provided from valve to raspbian not sure if they even work for OSMC
```bash
uname -a # check if your on armhf or aarch64
wget <link>
sudo dpkg <package>
```

**Comment by swetoast** [comment]
@briantogo @0chroma

**Comment by 0chroma** [comment]
tried it with the arm64 link, still same error :c

**Comment by swetoast** [comment]
@slouken did something change in the package update links after it got published to rasbian repo ?

**Comment by slouken** [comment]
Yes, there's a new format for update links to support multiple distributions:
```
OS_CODENAME=$(grep VERSION_CODENAME /etc/os-release | sed 's,.*=,,')
ARCH=$(dpkg --print-architecture)
UPDATE_URL="https://media.steampowered.com/steamlink/rpi/$OS_CODENAME/$ARCH"
```

**Comment by slouken** [comment]
What is the VERSION_CODENAME for OSMC? I can add that to the supported distributions.

**Comment by briantogo** [comment]
I tried both links swetoast, same error.  Steam finally did respond by sending me to their Steamlink page, which did not work.

**Comment by swetoast** [comment]
````bash
grep VERSION_CODENAME /etc/os-release | sed 's,.*=,,'
dpkg --print-architecture
````
@0chroma @briantogo  can anyone of you guys run these commands and retur the values please i dont have a rpi running osmc atm :+1:

**Comment by swetoast** [comment]
@samnazarko if you have a moment to chime in :+1:

**Comment by briantogo** [comment]
Hi Swetoast, is this what you're looking for?

osmc@osmc:~$ grep VERSION_CODENAME /etc/os-release | sed 's,.*=,,'
osmc@osmc:~$ dpkg --print-architecture
armhf
osmc@osmc:~$

**Comment by swetoast** [comment]
almost the first command didnt it produce anything ?

**Comment by slouken** [comment]
Can you include the full text of the file /etc/os-release?

**Comment by swetoast** [comment]
```bash
cat /etc/os-release
```

**Comment by briantogo** [comment]
osmc@osmc:~$ cat /etc/os-release
PRETTY_NAME="Open Source Media Center"
NAME="OSMC"
VERSION="October 2024"
VERSION_ID="2024.10-1"
ID=osmc
ID_LIKE=debian

ANSI_COLOR="1;31"
HOME_URL="https://osmc.tv"
SUPPORT_URL="https://osmc.tv"
BUG_REPORT_URL="https://osmc.tv"
osmc@osmc:~$

**Comment by slouken** [comment]
Can you see if `grep -E '^ID=' /etc/os-release | sed 's,.*=,,'` returns "osmc"?

**Comment by slouken** [comment]
The next question is, which version of Debian is OSMC most closely related to? bullseye or bookworm?

**Comment by swetoast** [comment]
```bash
cat /etc/apt/sources.list
```
should provide that info been awhile since i had OSMC running on any pi :)

**Comment by briantogo** [comment]
Yes it does return OSMC, and I'm ot sure which version.  However, the official OSMC site says the September release updated it to Bullseye, so guessing that may be my version too.

**Comment by briantogo** [comment]
Is there a command to remove any failed instalations of steamlink?  If I do anymore testing, it would be nice to get rid of old versions first.  I can't find them under add-ons in the OSMC gui.

**Comment by slouken** [comment]
Don't worry, you don't have anything broken to clean up, I think we can get you working here.

Can you run this and post the output?
`cat /etc/apt/sources.list`

**Comment by briantogo** [comment]
osmc@osmc:~$ cat /etc/apt/sources.list
deb https://deb.debian.org/debian bullseye main contrib non-free

deb https://deb.debian.org/debian/ bullseye-updates main contrib non-free

deb https://security.debian.org/ bullseye-security main contrib non-free

deb https://apt.osmc.tv bullseye main

osmc@osmc:~$

**Comment by slouken** [comment]
Can you apply this patch? If not, I can get a version of /usr/bin/steamlink for you in a little while.
```diff
--- /usr/bin/steamlink	2024-11-10 16:40:55.000000000 -0800
+++ /usr/bin/steamlink	2024-11-10 16:40:55.000000000 -0800
@@ -4,8 +4,15 @@
 # It will download the Steam Link files if necessary and launch them
 
 OS_CODENAME=$(grep VERSION_CODENAME /etc/os-release | sed 's,.*=,,')
-ARCH=$(dpkg --print-architecture)
-UPDATE_URL="https://media.steampowered.com/steamlink/rpi/$OS_CODENAME/$ARCH"
+if [ "$OS_CODENAME" = "" ]; then
+	if [ -f /etc/apt/sources.list ]; then
+		OS_CODENAME=$(grep 'deb.debian.org' /etc/apt/sources.list | head -1 | awk '{print $3}')
+	fi
+fi
+if [ "$OS_CODENAME" != "" ]; then
+	ARCH=$(dpkg --print-architecture)
+	UPDATE_URL="https://media.steampowered.com/steamlink/rpi/$OS_CODENAME/$ARCH"
+fi
 STEAMPACKAGE=steamlink
 STEAMLINK=steamlink.sh
 STEAMLINK_DATA_HOME="${XDG_DATA_HOME:-$HOME/.local/share}"/SteamLink
@@ -217,6 +224,8 @@
 # Launch the Steam Link app if possible
 if [ -x "$STEAMLINK_DATA_HOME/$STEAMLINK" ]; then
 	exec "$STEAMLINK_DATA_HOME/$STEAMLINK" "$@"
+elif [ "$OS_CODENAME" = "" ]; then
+	show_message --error $"Couldn't detect Debian version, aborting"
 else
 	show_message --error $"Couldn't download Steam Link application, aborting"
 fi
```

**Comment by briantogo** [comment]
Sorry Slouken, I don't know how to do that

**Comment by briantogo** [comment]
OK, so I added the green lines to the file /usr/bin/steamlink, but now I have the following errors;


osmc@osmc:~$ steamlink
/usr/bin/steamlink: line 1: ---: command not found
/usr/bin/steamlink: line 2: +++: command not found
/usr/bin/steamlink: line 3: @@: command not found
/usr/bin/steamlink: line 7: -ARCH=armhf: command not found
/usr/bin/steamlink: line 8: -UPDATE_URL=https://media.steampowered.com/steamlink/rpi//: No such file or directory
/usr/bin/steamlink: line 9: syntax error near unexpected token `then'
/usr/bin/steamlink: line 9: `+if [ "$OS_CODENAME" = "" ]; then'
osmc@osmc:~$ sudo nano /usr/bin/xsteamlink
osmc@osmc:~$ sudo nano /usr/bin/xsteamlink
osmc@osmc:~$ sudo nano /usr/bin/steamlink
osmc@osmc:~$ steamlink
/usr/bin/steamlink: line 6: ---: command not found
/usr/bin/steamlink: line 7: +++: command not found
/usr/bin/steamlink: line 8: @@: command not found
/usr/bin/steamlink: line 13: syntax error near unexpected token `then'
/usr/bin/steamlink: line 13: `+if [ "$OS_CODENAME" = "" ]; then'
osmc@osmc:~$

**Comment by samnazarko** [comment]
OSMC has been based on Bullseye since September 2022.
The September 2024 update did not update OSMC to Debian Bullseye

From: briantogo ***@***.***>
Reply to: swetoast/steamlink-launcher ***@***.***>
Date: Monday, 9 December 2024 at 16:58
To: swetoast/steamlink-launcher ***@***.***>
Cc: Sam Nazarko ***@***.***>, Mention ***@***.***>
Subject: Re: [swetoast/steamlink-launcher] Won't launch Steamlink (Issue #61)


Yes it does return OSMC, and I'm ot sure which version. However, the official OSMC site says the September release updated it to Bullseye, so guessing that may be my version too.

**Comment by swetoast** [comment]
the vero devices @samnazarko what would you need in order for those to work 

Display system?
Video accel system?

what is the vero running ?

**Comment by samnazarko** [comment]
GLES for windowing
AMCodec for video playback


On 10 Dec 2024, at 07:03, Toast ***@***.***> wrote:

﻿

the vero devices @samnazarko<https://github.com/samnazarko> what would you need in order for those to work

Display system?
Video accel system?

what is the vero running ?

**Comment by swetoast** [comment]
@slouken any possibility of adding support for that making it work on every OSMC device not just the raspberry pi ?

**Comment by slouken** [comment]
> OK, so I added the green lines to the file /usr/bin/steamlink, but now I have the following errors;

Try downloading and running this:
```
wget https://www.libsdl.org/tmp/steamlink
chmod 755 steamlink
sudo cp steamlink /usr/bin
steamlink
```

**Comment by slouken** [comment]
> @slouken any possibility of adding support for that making it work on every OSMC device not just the raspberry pi ?

If it's Debian based, the steamlink at https://www.libsdl.org/tmp/steamlink should at least let you try, but there are no guarantees.

**Comment by briantogo** [comment]
I gave it a try Slouken, but it's stuck on "No Carrier"

osmc@osmc:~$ steamlink
You are running with less than 128 MB video memory, you may need to go to the Raspberry Pi Configuration and increase your GPU memory.
Press enter to continue:
Failed to move cursor on screen HDMI1: -13
Could not set cursor on screen HDMI1: -13
Warning: Couldn't set IPV6_ADD_MEMBERSHIP on discovery network socket
"No carrier"
Mix_OpenAudioDevice() failed: No default audio device available
Couldn't find sound data in ":/resources/sounds/click.wav"
Failed to move cursor on screen HDMI1: -13
Could not set cursor on screen HDMI1: -13
Could not set cursor on screen HDMI1: -13
Connected to Remote Client service with session 2809238279746730511
Could not set DRM mode for screen HDMI1 (Permission denied)
Could not queue DRM page flip on screen HDMI1 (Permission denied)
Could not queue DRM page flip on screen HDMI1 (Permission denied)
Could not queue DRM page flip on screen HDMI1 (Permission denied)
"No carrier"
"No carrier"

**Comment by slouken** [comment]
I haven't seen that message before, what's your network setup?

**Comment by briantogo** [comment]
Xplornet in Canada, TP-Link mesh router.  The pi is connected via ethernet.  First I've seen this message too since I tried the new version of Steamlink

**Comment by slouken** [comment]
>  Could not set DRM mode for screen HDMI1 (Permission denied).
>  Could not queue DRM page flip on screen HDMI1 (Permission denied) 

It looks like this is the real problem. It seems like the OS isn't configured for user access to DRM resources.

**Comment by samnazarko** [comment]
We are not using DRM yet on Vero. But we are on RPi.

I suspect video acceleration also uses GBM for windowing and v4l2.

On 10 Dec 2024, at 23:18, Sam Lantinga ***@***.***> wrote:

﻿

Could not set DRM mode for screen HDMI1 (Permission denied).
Could not queue DRM page flip on screen HDMI1 (Permission denied)

It looks like this is the real problem. It seems like the OS isn't configured for user access to DRM resources.

**Comment by briantogo** [comment]
Just to clarify, I am using a Raspberry Pi 4 running OSMC

**Comment by samnazarko** [comment]
That uses DRM so shouldn’t be a problem,

On 10 Dec 2024, at 23:54, briantogo ***@***.***> wrote:

﻿

Just to clarify, I am using a Raspberry Pi 4 running OSMC

**Comment by briantogo** [comment]
I tried it again today Slouken, same result

**Comment by briantogo** [comment]
Finally heard from Steam on this issue;

Hi there,

I'm sorry that you're encountering an issue with Remote Play via Steam Link.

Due to the number of unique machine configurations, network setups, firewall settings, and various software configurations, we don't provide in-depth technical support for Remote Play.

Please visit our [Wireless Streaming FAQ](https://support.steampowered.com/kb_article.php?ref=4950-EBNM-7843) for recommendations on improving your performance. Then, please go through our [Troubleshooting Network Connectivity](https://support.steampowered.com/kb_article.php?ref=1456-EUDN-2493) article for probable fixes even if it seems like your network is functioning correctly.

We also encourage you to look through our [Remote Play FAQ](https://support.steampowered.com/kb_article.php?ref=3629-RIAV-1617) which includes the most common technical issues that players encounter along with a basic troubleshooting guide.

As an additional resource, we encourage players to post in the [Steam Discussions](https://steamcommunity.com/groups/homestream/discussions/) where other players may have had success resolving similar issues. If you are able to reproduce a bug with Remote Play, please post a detailed description of the bug along with the steps that can be taken to reproduce it within these forums.

I am sorry that we don't have a specific fix for your issue.

Steam Support
Lucio

**Comment by slouken** [comment]
I'm not sure what's happening here. OSMC isn't officially supported. You might try asking on some OSMC forums, and see if anyone has any suggestions to get more info. I suspect it's a permissions issue, but Bullseye never worked very well and required hacks like setting up fkms, so I wouldn't hold my breath.

**Comment by swetoast** [comment]
no need to go to OSMC forums since @samnazarko the main dev is here :)

**Comment by rootabeta** [comment]
I've identified the root cause of the issue - the steam link install script is grepping /etc/os-release for the VERSION_CODENAME to fetch the steam link file from Steam servers. However, OSMC does not ship with VERSION_CODENAME set, nevermind set to a value Steam has a directory for, so the request fails. Using the URL provided by swetoast, I modified the bash file at /usr/bin/steamlink by injecting the following above line 188
```diff
  pushd "$STEAMLINK_TMPDIR" >/dev/null
  archive=$(basename "$latest")
+ latest="https://media.steampowered.com/steamlink/rpi/latest/steamlink.deb"
  if download_archive "Downloading update" "$latest"; then```

**Comment by new-2-this** [comment]
> > OK, so I added the green lines to the file /usr/bin/steamlink, but now I have the following errors;
> 
> Try downloading and running this:
> 
> ```
> wget https://www.libsdl.org/tmp/steamlink
> chmod 755 steamlink
> sudo cp steamlink /usr/bin
> steamlink
> ```

This worked for me on Ri3b running OSMC. 

128MB GPU MEM, still got warning that it is less than 128MB. Used .ignore_gpumem to get around that. Would be great if the script asked to create that file. More information on how to extend the memory would also be appreciated. Seems people on the internet had different ideas. Anyway, great project and it really saved me today. Thank you guys!

**Comment by Pieeeeeee** [comment]
> > OK, so I added the green lines to the file /usr/bin/steamlink, but now I have the following errors;
> 
> Try downloading and running this:
> 
> ```
> wget https://www.libsdl.org/tmp/steamlink
> chmod 755 steamlink
> sudo cp steamlink /usr/bin
> steamlink
> ```
I'm having the same problem as OP on a Vero 4k+. Is there something else you need to do besides this script? I get an error on the last line, saying "Couldn't detect Debian version, aborting".

**Comment by swetoast** [comment]
vero isnt supported
